templates/raspbian/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers

```diff
@@ -124,6 +124,61 @@ cattlepi_fetch_update_config()
     cattlepi_fetch_update_config_invoked=yes
 }
 
+# apply the requested sd disk layout
+cattlepi_check_apply_disk_layout()
+{
+    cattlepi_fetch_update_config
+    jq -r ".config.sdlayout" "${CATTLEPI_DIR}/config" | grep -q null
+    if [ $? -ne 0 ]; then
+        # pull down and decode the requested disk layout
+        jq -r ".config.sdlayout" "${CATTLEPI_DIR}/config" | base64 -d > /tmp/sdlayout
+        # TODO: there should be proper error checking here, but do the bare minimum here
```

Comment:
Yes I agree - especially when dealing with disks! :)

Most of these calls could benefit from a function that checks the exit code and exits if things go bad.  And speaking of things going bad - if things really go south it seems as though there should be a fallback to putting the initfs back where it was and booting back into cattle pi.  

Perhaps there should be a flag somewhere or persistent logging location so a user could learn what happend?

Reply:
I like the idea of a "safe_mode" where if things go wrong we at least leave it in a state that will try again on the next boot. most of the code was written w/ that in mind but we would need extensive testing to prove it. 

There is an API we could use for logging: POST https://api.cattlepi.com/track/{deviceid} See https://cattlepi.com/api/ Right now it's pretty limited but it could be evolved. We could also develop something specifically for logs.

Reply:
I like either idea (API or something specific for logs)!